include/flatbuffers/flatbuffers.h

```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
   }
 };
 
-// Simple indirection for buffer allocation, to allow this to be overridden
-// with custom allocation (see the FlatBufferBuilder constructor).
-class simple_allocator {
+// Allocator interface. This is flatbuffers-specific and meant only for
+// `vector_downward` usage. In particular, one important difference from a
+// "normal" allocator is that there is only a single outstanding memory region
+// at a time. This is because the allocator is owned by a single
+// `vector_downward` and it only ever has one region active.
+class Allocator {
  public:
-  virtual ~simple_allocator() {}
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  virtual ~Allocator() {}
+
+  // Allocate `size` bytes of memory.
+  virtual uint8_t *allocate(size_t size) = 0;
+
+  // Deallocate `size` bytes of memory at `p` allocated by this allocator.
+  virtual void deallocate(uint8_t *p, size_t size) = 0;
+
+  // Reallocate `new_size` bytes of memory, replacing the old region of size
+  // `old_size` at `p`. In contrast to a normal realloc, this grows downwards,
+  // and is intended specifcally for `vector_downward` use.
+  virtual uint8_t *reallocate_downward(
```

Comment:
`reallocate_downward` is actually overridden in the gRPC allocator, which is why I added this method. By letting the allocator handle the memcpy, the allocator no longer needs to have up to 2 allocate regions outstanding, and can just do the temp work in the function scope while it's reallocating. If you compare that PR's old/new impl you'll see that it's simplified significantly.